src/Features/Core/Portable/Diagnostics/DiagnosticIdFilter.cs

```diff
@@ -0,0 +1,51 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+
+namespace Microsoft.CodeAnalysis.Diagnostics;
+
+/// <param name="IncludedDiagnosticIds">If present, if an analyzer has at least one descriptor in this set, it will be included.</param>
+/// <param name="ExcludedDiagnosticIds">If present, if all of the descriptors an analyzer has is in this set, it will be excluded.</param>
+internal readonly record struct DiagnosticIdFilter(
+    ImmutableHashSet<string>? IncludedDiagnosticIds,
+    ImmutableHashSet<string>? ExcludedDiagnosticIds)
+{
+    public static readonly DiagnosticIdFilter All = default;
+
+    public static DiagnosticIdFilter Include(ImmutableHashSet<string>? includedDiagnosticIds)
+        => new(includedDiagnosticIds, ExcludedDiagnosticIds: null);
+
+    public static DiagnosticIdFilter Exclude(ImmutableHashSet<string> excludedDiagnosticIds)
+        => new(IncludedDiagnosticIds: null, excludedDiagnosticIds);
+
+    public bool Allow(params IEnumerable<string> ids)
+    {
+        if (this == All)
+            return true;
+
+        foreach (var id in ids)
+        {
+            // If the ID is in the included set, then that's good enough as the semantics for this type are that as long
+            // as we see one allowed id, we allow the analyzer.
+            if (IncludedDiagnosticIds != null &&
+                IncludedDiagnosticIds.Contains(id))
+            {
+                return true;
+            }
+
+            // If the ID is *not* in the excluded set, then that's good enough as the semantics for this type are that we
+            // only filter out the analyzers that have *all* their IDs in the excluded set.
```

Comment:
I think this method is a little confusing - is `ids` always associated with a single analyzer?  Or can it contain ids from different analyzers?

If just a single one, I'd rename it to make it more clear that this is only expected to be called with `ids` from a single analyzer.  That also makes the comment here make more sense (why the semantics here relate to `analyzers` and not just `ids`).

Reply:
>  is ids always associated with a single analyzer? Or can it contain ids from different analyzers?

It can contain ids from many analyzers.  A caller can basically say "include an anlyzer if it contains at least one id from this whitelisted set".  Or it can alternatively says "blacklist an analyzer if all of its ids are on this disallowed group".

Note: the last one is weird, and only used in one location.  It's during the "run code cleanup" pass when we are explicitly only running 3rd party analyzers.  we accomplish this by putting all our own roslyn analyzer supported IDs on the blacklist.  that way our own analyzers are blocked and we only run other analyzers.